backend/src/org/commcare/session/SessionNavigator.java

```diff
@@ -135,7 +135,8 @@ private void handleCompute() {
             currentSession.setComputedDatum(ec);
         } catch (XPathException e) {
             this.thrownException = e;
-            sendResponse(EXCEPTION_THROWN);
+            sendResponse(XPATH_EXCEPTION_THROWN);
+            return;
```

Comment:
The absence of this return statement is what caused the `XPathException` that was being thrown to manifest itself as an infinite loop and eventual force close.